lexis/render/ColorMap.h

```diff
@@ -0,0 +1,249 @@
+     * @returns colors that are interpolated for given control points range. If there are no
+     * control points for a channel, the returned colors will have 0 values for that channel.
+     * If there is only one control point, all samples will have the same color for that channel.
+     */
+    template< typename T >
+    LEXIS_API Colors< T > sampleColors( size_t count, float min, float max, T emptyColor );
+
+    /**
+     * Samples colors and fills the given buffer linearly in the given region of control points,
+     * given the number of sample points
+     * @param data is filled with the color values in RGBARGBA.... form. If there are no
+     * control points for a channel, the returned colors will have 0 values for that channel.
+     * If there is only one control point, all samples will have the same color for that channel.
+     * @param count number of sample points
+     * @param min minimum of the range to sample
+     * @param max maximum of the range to sample
+     * @param emptyColor the samples which are outside the boundaries filled with emptyColor
+     * @returns colors that are interpolated for given control points range.
+     */
+    template< typename T >
+    LEXIS_API void sampleColors( T* data, size_t count, float min, float max, T emptyColor );
+
+    /**
+     * Adds a control point to a given channel. If there is already a control point for the
+     * given cp.x, it is replaced with the given control point.
+     * @param cp control points
+     */
+    LEXIS_API void addControlPoint( const detail::ControlPoint& cp, Channel channel );
+
+    /** returns true if there are no control points in any channel */
+    LEXIS_API bool isEmpty() const;
+
+private:
+
+    ::zerobuf::Vector< detail::ControlPoint >& _getControlPoints( Channel channel );
+    void _sortChannels();
+    bool _isSorted;
+};
+
+template< typename T >
+Colors< T > ColorMap::sampleColors( size_t count )
+{
+    return sampleColors< T >( count,
+                              -std::numeric_limits< float >::infinity(),
+                              std::numeric_limits< float >::infinity(),
+                              static_cast< T >( 0 ));
+}
+
+template< typename T >
+void ColorMap::sampleColors( T* data, size_t count )
+{
+    return sampleColors< T >( data,
+                              count,
+                              -std::numeric_limits< float >::infinity(),
+                              std::numeric_limits< float >::infinity(),
+                              static_cast< T >( 0 ));
+}
+
+template< typename T >
+Colors< T > ColorMap::sampleColors( size_t count, float min, float max, T emptyColor )
+{
+    Colors< T > colors( count, { emptyColor, emptyColor, emptyColor, emptyColor });
+    sampleColors( reinterpret_cast< T* >( colors.data( )), count, min, max, emptyColor );
+    return colors;
+}
+
+template< typename T >
+void ColorMap::sampleColors( T* data, size_t count, float min, float max, T emptyColor )
```

Comment:
Maybe you can move the implementation to a cpp file and instantiate the template for unsigned char and float there (I don't think we need any other type apart from those). The rest will cause a link error, but can be easily added if needed.